src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs

```diff
@@ -304,8 +304,12 @@ public bool ExecuteReplicationOnce(TcpConnectionOptions tcpConnectionOptions, Ou
                             }
 
                             replicationState.Size += item.Size;
-
                             replicationState.NumberOfItemsSent++;
+
+                            if (replicationState.Size >= _parent._database.Configuration.Replication.MaxSizeToLoadFromStorage?.GetValue(SizeUnit.Bytes))
```

Comment:
If the root cause is indeed that we skip too much.
This condition seems to be in the wrong place... we should break the tx _before_ we can even send anything

Reply:
You want to send, but you want to stop loading items from storage to the list ```_orderedReplicaItems``` that causing OOM because it holds too much data in it.
you load it in ```AddReplicationItemToBatch(item, _stats.Storage, replicationState, skippedReplicationItemsInfo)``` inside the loop.

Reply:
That isn't the problem, however. 
It isn't that the `_orderedReplicaItems` is too big. The problem is that we are _skipping_ too many items, but the encrypted data is stuck in memory.

We need to take that into account when figuring out when to break the batch, see:

https://github.com/ravendb/ravendb/blob/b7342a4a4cb53d0c7e5b039759cf24c8da9fb5aa/src/Raven.Server/Documents/Indexes/Index.cs#L4377

Reply:
Yes I see, I fixed it in different way